### PR TITLE
8344526: RISC-V: implement -XX:+VerifyActivationFrameSize

### PR DESCRIPTION
**Pull request by sendaoYan**
Hi all,
Currently on linux-riscv64 platform the debug VM option `-XX:+VerifyActivationFrameSize` is Unimplemented. I want to implement `-XX:+VerifyActivationFrameSize` to make JVM easier to diagnose. Only effect debug build plus with option `-XX:+VerifyActivationFrameSize`, the change has been verified locally, the risk is low.

Additional testing

- [x] Run SPECjbb2015 with -XX:+VerifyActivationFrameSize option

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344526](https://bugs.openjdk.org/browse/JDK-8344526): RISC-V: implement -XX:+VerifyActivationFrameSize (**Enhancement** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) Review applies to [bc90d67d](https://git.openjdk.org/jdk/pull/22264/files/bc90d67d0050b14a85b943a2f458721b5b1fcd06)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Contributors
 * Fei Yang `<fyang@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22264/head:pull/22264` \
`$ git checkout pull/22264`

Update a local copy of the PR: \
`$ git checkout pull/22264` \
`$ git pull https://git.openjdk.org/jdk.git pull/22264/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22264`

View PR using the GUI difftool: \
`$ git pr show -t 22264`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22264.diff">https://git.openjdk.org/jdk/pull/22264.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22264#issuecomment-2487253920)
</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 1]
```diff
@@ -441,7 +441,13 @@ void InterpreterMacroAssembler::dispatch_base(TosState state,
                                               Register Rs) {
   // Pay attention to the argument Rs, which is acquiesce in t0.
   if (VerifyActivationFrameSize) {
-    Unimplemented();
+    Label L;
+    sub(t1, fp, esp);
+    int min_frame_size = (frame::link_offset - frame::interpreter_frame_initial_sp_offset) * wordSize;
+    add(t1, t1, -min_frame_size);
```
[the remark this answers is not in the document]
Comment (sendaoYan) [reply src/hotspot/cpu/riscv/interp_masm_riscv.cpp]: Thanks, the `add` has been replaced as `sub`

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 2]
```diff
@@ -441,7 +441,13 @@ void InterpreterMacroAssembler::dispatch_base(TosState state,
                                               Register Rs) {
   // Pay attention to the argument Rs, which is acquiesce in t0.
   if (VerifyActivationFrameSize) {
-    Unimplemented();
+    Label L;
+    sub(t1, fp, esp);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:445]: Note that RISC-V is a bit different in frame structure from other CPU platforms. I think we should exlcude the 2 frame metadata words after this `sub`, like:
```
sub(t1, t1, frame::metadata_words * wordSize); // Exclude 2 frame metadata words
```

Comment (sendaoYan) [reply src/hotspot/cpu/riscv/interp_masm_riscv.cpp:445]: Thanks, I found that `sp -esp` equals 96, larger than `min_frame_size` 16. I think the extra 16 bytes is the `2 frame metadata words`.

**Review by RealFYang — APPROVED** [review]

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 3]
```diff
@@ -441,7 +441,14 @@ void InterpreterMacroAssembler::dispatch_base(TosState state,
...
+    sub(t1, t1, frame::metadata_words * wordSize); // Exclude 2 frame metadata words
+    int min_frame_size = (frame::link_offset - frame::interpreter_frame_initial_sp_offset) * wordSize;
```
Comment (Hamlin-Li) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:447]: Could we also remove the above `sub`, and add `frame::metadata_words * wordSize` in this statement? it's not necessary to calculate the value in the generated code and seems to me the code will be more straight.

**Review by Hamlin-Li — COMMENTED** [review]
Looks good, just one minor comment.

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp (hunk 3 above)
Comment (RealFYang) [reply src/hotspot/cpu/riscv/interp_masm_riscv.cpp:447]: Yeah, I agree that will be cleaner.
```
int min_frame_size = (frame::link_offset - frame::interpreter_frame_initial_sp_offset + frame::metadata_words) * wordSize;
```

Comment (sendaoYan) [reply src/hotspot/cpu/riscv/interp_masm_riscv.cpp]: Thanks, the second `sub` calculation has been merged to `min_frame_size ` statement.

**Review by Hamlin-Li — APPROVED** [review]
Looks good, Thanks!

**Comment by sendaoYan** [comment]
Thanks all for the reviews.

/integrate

**Pull request closed without merging**